ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java

```diff
@@ -288,15 +313,231 @@ private void acquireLocksInternal() throws CommandProcessorException, LockExcept
+
+    for (Map.Entry<String, Table> tableInfo : writtenTables.entrySet()) {
+      String fullQNameForLock = TableName.getDbTable(tableInfo.getValue().getDbName(),
+          MetaStoreUtils.encodeTableName(tableInfo.getValue().getTableName()));
+      if (nonSharedLockedTables.contains(fullQNameForLock)) {
+        // Check if table is transactional
+        if (AcidUtils.isTransactionalTable(tableInfo.getValue())) {
+          ValidWriteIdList writeIdList = txnWriteIdList.getTableValidWriteIdList(tableInfo.getKey());
+          ValidWriteIdList currentWriteIdList = currentTxnWriteIds.getTableValidWriteIdList(tableInfo.getKey());
+          // Check if there was a conflicting write between current SNAPSHOT generation and locking.
+          if (currentWriteIdList.isWriteIdRangeValid(writeIdList.getHighWatermark() + 1,
+              currentWriteIdList.getHighWatermark()) != ValidWriteIdList.RangeResponse.NONE) {
+            return false;
+          }
+          // Check that write id is still valid
+          if (!TxnIdUtils.checkEquivalentWriteIds(writeIdList, currentWriteIdList)) {
+            // Write id has changed, it is not valid anymore, we need to recompile
+            return false;
+          }
+        }
+        nonSharedLockedTables.remove(fullQNameForLock);
+      }
+    }
+
+    if (!nonSharedLockedTables.isEmpty()) {
+      throw new LockException("Wrong state: non-shared locks contain information for tables that have not" +
+          " been visited when trying to validate the locks from query tables.\n" +
+          "Tables: " + writtenTables.keySet() + "\n" +
+          "Remaining locks after check: " + nonSharedLockedTables);
+    }
+
+    return true; // It passes the test, it is valid
+  }
+
+  private Map<String, Table> getTables(boolean inputNeeded, boolean outputNeeded) {
+    Map<String, Table> tables = new HashMap<>();
+    if (inputNeeded) {
+      driverContext.getPlan().getInputs().forEach(input -> addTableFromEntity(input, tables));
+    }
+    if (outputNeeded) {
+      driverContext.getPlan().getOutputs().forEach(output -> addTableFromEntity(output, tables));
+    }
+    return tables;
+  }
+
+  private void addTableFromEntity(Entity entity, Map<String, Table> tables) {
+    Table table;
+    switch (entity.getType()) {
+    case TABLE:
+      table = entity.getTable();
+      break;
+    case PARTITION:
+    case DUMMYPARTITION:
+      table = entity.getPartition().getTable();
+      break;
+    default:
+      return;
+    }
+    String fullTableName = AcidUtils.getFullTableName(table.getDbName(), table.getTableName());
+    tables.put(fullTableName, table);
+  }
+
+  private List<String> getTransactionalTables(Map<String, Table> tables) {
+    return tables.entrySet().stream()
+      .filter(entry -> AcidUtils.isTransactionalTable(entry.getValue()))
+      .map(Map.Entry::getKey)
+      .collect(Collectors.toList());
+  }
+
+  void addHiveLocksFromContext() {
```

Comment:
I do not like storing lock related stuff in context. Shouldn't this be private to the DbTxnHandler?

Reply:
I agree, but in a separate jira. This one is for merging the two classes only, so we'll have a clean and easy to understand history.